sdk/data/azcosmos/cosmos_global_endpoint_manager.go

```diff
@@ -0,0 +1,186 @@
+package azcosmos
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	azlog "github.com/Azure/azure-sdk-for-go/sdk/azcore/log"
+	azruntime "github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+	"github.com/Azure/azure-sdk-for-go/sdk/internal/log"
+)
+
+const globalEndpointManagerDefaultBackgroundRefreshInterval = 5 * time.Minute
+
+type globalEndpointManager struct {
+	client           *Client
+	preferredRegions []string
+	locationCache    *locationCache
+	refreshTicker    *time.Ticker
+	refreshDone      chan struct{}
+	sync.Mutex
+}
+
+// NewGlobalEndpointManager creates a new global endpoint manager.
+func newGlobalEndpointManager(client *Client, preferredRegions []string, refreshInterval time.Duration) (*globalEndpointManager, error) {
+	endpoint, err := url.Parse(client.endpoint)
+	if err != nil {
+		return nil, err
+	}
+	gem, err := &globalEndpointManager{
+		client:           client,
+		preferredRegions: preferredRegions,
+		locationCache:    newLocationCache(preferredRegions, *endpoint),
+	}, nil
+	if err != nil {
+		return nil, err
+	}
+	gem.stopBackgroundRefresh()
+	gem.startBackgroundRefresh(refreshInterval)
+	defer gem.stopBackgroundRefresh()
+
+	return gem, nil
+}
+
+// GetWriteEndpoints returns the write endpoints from the location cache.
+func (gem *globalEndpointManager) GetWriteEndpoints() ([]url.URL, error) {
+	return gem.locationCache.writeEndpoints()
+}
+
+// GetReadEndpoints returns the read endpoints from the location cache.
+func (gem *globalEndpointManager) GetReadEndpoints() ([]url.URL, error) {
+	return gem.locationCache.readEndpoints()
+}
+
+// GetAccountProperties retrieves account properties from the Cosmos DB instance.
+func (gem *globalEndpointManager) GetAccountProperties() (accountProperties, error) {
+	operationContext := pipelineRequestOptions{
+		resourceType:    resourceTypeDatabaseAccount,
+		resourceAddress: "",
+	}
+
+	path, err := generatePathForNameBased(resourceTypeDatabaseAccount, "", false)
+	if err != nil {
+		return accountProperties{}, fmt.Errorf("failed to generate path for name-based request: %w", err)
```

Comment:
Using `%w` will wrap the error, meaning it can be unwrapped via `errors.Is()` and/or `errors.As()`.  As a result, the wrapped error is publicly accessible which is usually undesirable.  It would be better to use `%v` instead, or just remove the `failed to generate...` text altogether and jus return `err`.

Comment applies to all places where `%w` is used.